File: sentry-replays-meticulous/replay_analyzer.py
#!/usr/bin/env python3
"""
Replay Analyzer

Analyzes Sentry replay data to extract patterns suitable for automated testing.
This serves as a bridge between Sentry's replay format and a meticulous.ai-style
testing system.
"""

import json
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, asdict
from datetime import datetime
from collections import Counter


@dataclass
class UserAction:
    """Represents a single user action extracted from replay."""
    timestamp: int  # milliseconds since replay start
    action_type: str  # click, input, navigation, etc.
    selector: Optional[str] = None  # CSS selector or XPath
    value: Optional[str] = None  # For input events
    target: Optional[str] = None  # URL for navigation
    coordinates: Optional[Dict[str, int]] = None  # x, y for clicks


@dataclass
class NetworkRequest:
    """Represents a network request during replay."""
    timestamp: int
    method: str
    url: str
    status_code: Optional[int] = None
    response_body: Optional[str] = None
    request_body: Optional[str] = None


@dataclass
class ReplaySession:
    """Analyzed replay session suitable for automated testing."""
    replay_id: str
    duration: int  # milliseconds
    url: str  # Starting URL
    actions: List[UserAction]
    network_requests: List[NetworkRequest]
    errors: List[Dict[str, Any]]
    metadata: Dict[str, Any]

    def to_test_script(self) -> str:
        """Generate a Playwright test script from this session."""
        script_lines = [
            "import { test, expect } from '@playwright/test';",
            "",
            f"test('replay-{self.replay_id[:8]}', async ({{ page }}) => {{",
            f"  // Original session duration: {self.duration}ms",
            f"  // URL: {self.url}",
            "",
            f"  await page.goto('{self.url}');",
            ""
        ]

        for action in self.actions:
            if action.action_type == 'click' and action.selector:
                script_lines.append(f"  await page.click('{action.selector}');")
            elif action.action_type == 'input' and action.selector and action.value:
                script_lines.append(f"  await page.fill('{action.selector}', '{action.value}');")
            elif action.action_type == 'navigation' and action.target:
                script_lines.append(f"  await page.goto('{action.target}');")

            # Add small delay to simulate real user behavior
            script_lines.append("  await page.waitForTimeout(100);")
            script_lines.append("")

        script_lines.append("  // Take final screenshot for comparison")
        script_lines.append(f"  await page.screenshot({{ path: 'screenshots/replay-{self.replay_id[:8]}.png' }});")
        script_lines.append("});")

        return "\n".join(script_lines)


class ReplayAnalyzer:
    """Analyzes Sentry replays to extract testable patterns."""

    def __init__(self):
        self.sessions: List[ReplaySession] = []

    def analyze_replay(self, replay_data: Dict[str, Any]) -> ReplaySession:
        """
        Analyze a single replay and extract actionable test data.

        Args:
            replay_data: Raw replay data from Sentry API

        Returns:
            Analyzed ReplaySession object
        """
        replay_id = replay_data.get('id', 'unknown')
        duration = replay_data.get('duration', 0)

        # Extract starting URL (may be in various fields)
        url = self._extract_url(replay_data)

        # Parse events (if available in the replay data)
        # Note: Actual event structure depends on Sentry's replay format
        actions = self._extract_actions(replay_data)
        network_requests = self._extract_network_requests(replay_data)
        errors = self._extract_errors(replay_data)

        metadata = {
            'user': replay_data.get('user', {}),
            'browser': replay_data.get('browser', {}),
            'os': replay_data.get('os', {}),
            'device': replay_data.get('device', {}),
            'started_at': replay_data.get('started_at'),
            'finished_at': replay_data.get('finished_at'),
            'error_count': replay_data.get('error_count', 0),
            'project_id': replay_data.get('project_id')
        }

        session = ReplaySession(
            replay_id=replay_id,
            duration=duration,
            url=url,
            actions=actions,
            network_requests=network_requests,
            errors=errors,
            metadata=metadata
        )

        self.sessions.append(session)
        return session

    def _extract_url(self, replay_data: Dict[str, Any]) -> str:
        """Extract starting URL from replay data."""
        # Check various possible fields
        if 'urls' in replay_data and replay_data['urls']:
            return replay_data['urls'][0]
        if 'url' in replay_data:
            return replay_data['url']
        if 'tags' in replay_data:
            for tag in replay_data['tags']:
                if tag.get('key') == 'url':
                    return tag.get('value', '')

        return 'https://example.com'  # Fallback

    def _extract_actions(self, replay_data: Dict[str, Any]) -> List[UserAction]:
        """
        Extract user actions from replay data.

        Note: This is a placeholder. Actual implementation depends on
        Sentry's replay event format, which may require fetching
        additional data via the events endpoint.
        """
        actions = []

        # Placeholder: In a real implementation, you'd parse the
        # rrweb events or Sentry's specific format
        # Example structure we might see:
        # {
        #   "type": 3,  # IncrementalSnapshot
        #   "data": {
        #     "source": 2,  # MouseInteraction
        #     "type": 2,  # Click
        #     "id": 123,
        #     "x": 100,
        #     "y": 200
        #   }
        # }

        return actions

    def _extract_network_requests(self, replay_data: Dict[str, Any]) -> List[NetworkRequest]:
        """Extract network requests from replay data."""
        requests = []

        # Placeholder: Parse network activity from replay
        # Sentry may include this in breadcrumbs or separate events

        return requests

    def _extract_errors(self, replay_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Extract errors from replay data."""
        errors = []

        if 'errors' in replay_data:
            errors = replay_data['errors']
        elif 'error_ids' in replay_data:
            # Would need to fetch error details separately
            errors = [{'id': eid} for eid in replay_data['error_ids']]

        return errors

    def analyze_coverage(self) -> Dict[str, Any]:
        """
        Analyze test coverage across all sessions.

        Returns:
            Coverage statistics
        """
        all_urls = []
        all_selectors = []
        all_browsers = []

        for session in self.sessions:
            all_urls.append(session.url)
            all_browsers.append(session.metadata.get('browser', {}).get('name'))

            for action in session.actions:
                if action.selector:
                    all_selectors.append(action.selector)

        return {
            'total_sessions': len(self.sessions),
            'unique_urls': len(set(all_urls)),
            'unique_selectors': len(set(all_selectors)),
            'url_distribution': dict(Counter(all_urls)),
            'browser_distribution': dict(Counter(all_browsers)),
            'total_actions': sum(len(s.actions) for s in self.sessions),
            'total_errors': sum(len(s.errors) for s in self.sessions)
        }

    def select_sessions_for_testing(
        self,
        max_sessions: int = 10,
        prioritize_errors: bool = True,
        min_duration: int = 1000  # milliseconds
    ) -> List[ReplaySession]:
        """
        Select most valuable sessions for automated testing.

        Args:
            max_sessions: Maximum number of sessions to select
            prioritize_errors: Prioritize sessions with errors
            min_duration: Minimum session duration in ms

        Returns:
            List of selected sessions
        """
        # Filter by minimum duration
        candidates = [s for s in self.sessions if s.duration >= min_duration]

        if prioritize_errors:
            # Sort by error count (descending) then by duration
            candidates.sort(
                key=lambda s: (len(s.errors), s.duration),
                reverse=True
            )
        else:
            # Sort by duration (longer sessions = more coverage)
            candidates.sort(key=lambda s: s.duration, reverse=True)

        return candidates[:max_sessions]

    def export_for_testing(self, output_dir: str = './tests'):
        """
        Export selected sessions as Playwright test scripts.

        Args:
            output_dir: Directory to write test files
        """
        import os

        os.makedirs(output_dir, exist_ok=True)
        selected = self.select_sessions_for_testing()

        for i, session in enumerate(selected):
            test_script = session.to_test_script()
            filename = f"{output_dir}/replay-{session.replay_id[:8]}.spec.ts"

            with open(filename, 'w') as f:
                f.write(test_script)

            print(f"Generated test: {filename}")

        # Generate a summary file
        summary = {
            'generated_at': datetime.utcnow().isoformat(),
            'total_tests': len(selected),
            'coverage': self.analyze_coverage(),
            'tests': [
                {
                    'replay_id': s.replay_id,
                    'duration': s.duration,
                    'url': s.url,
                    'action_count': len(s.actions),
                    'error_count': len(s.errors)
                }
                for s in selected
            ]
        }

        with open(f"{output_dir}/test-summary.json", 'w') as f:
            json.dump(summary, f, indent=2)

        print(f"\nGenerated {len(selected)} tests in {output_dir}/")
        print(f"Summary written to {output_dir}/test-summary.json")


def main():
    """Example usage of ReplayAnalyzer."""

    # Example replay data (simplified)
    sample_replay = {
        'id': 'abc123def456',
        'duration': 45000,  # 45 seconds
        'url': 'https://example.com/app',
        'user': {'email': 'user@example.com'},
        'browser': {'name': 'Chrome', 'version': '120.0'},
        'os': {'name': 'macOS', 'version': '14.0'},
        'error_count': 2,
        'errors': [
            {'message': 'TypeError: Cannot read property X'},
            {'message': 'NetworkError: Failed to fetch'}
        ]
    }

    analyzer = ReplayAnalyzer()

    # Analyze replay
    session = analyzer.analyze_replay(sample_replay)

    print("Analyzed Replay Session:")
    print(f"  ID: {session.replay_id}")
    print(f"  Duration: {session.duration}ms")
    print(f"  URL: {session.url}")
    print(f"  Errors: {len(session.errors)}")

    # Generate test script
    print("\nGenerated Test Script:")
    print(session.to_test_script())

    # Coverage analysis
    print("\nCoverage Analysis:")
    coverage = analyzer.analyze_coverage()
    print(json.dumps(coverage, indent=2))


if __name__ == '__main__':
    main()

File: sentry-replays-meticulous/src/replay-analyzer.ts
/**
 * Replay Analyzer
 *
 * Analyzes Sentry replay data to extract patterns suitable for automated testing.
 * This serves as a bridge between Sentry's replay format and a meticulous.ai-style
 * testing system.
 */

import * as fs from 'fs';
import * as path from 'path';

export interface UserAction {
  timestamp: number; // milliseconds since replay start
  actionType: string; // click, input, navigation, etc.
  selector?: string; // CSS selector or XPath
  value?: string; // For input events
  target?: string; // URL for navigation
  coordinates?: { x: number; y: number }; // x, y for clicks
}

export interface NetworkRequest {
  timestamp: number;
  method: string;
  url: string;
  statusCode?: number;
  responseBody?: string;
  requestBody?: string;
}

export interface ReplayMetadata {
  user?: Record<string, any>;
  browser?: { name: string; version: string };
  os?: { name: string; version: string };
  device?: Record<string, any>;
  startedAt?: string;
  finishedAt?: string;
  errorCount: number;
  projectId?: number;
}

export interface ReplaySession {
  replayId: string;
  duration: number; // milliseconds
  url: string; // Starting URL
  actions: UserAction[];
  networkRequests: NetworkRequest[];
  errors: Array<Record<string, any>>;
  metadata: ReplayMetadata;
}

export interface CoverageStats {
  totalSessions: number;
  uniqueUrls: number;
  uniqueSelectors: number;
  urlDistribution: Record<string, number>;
  browserDistribution: Record<string, number>;
  totalActions: number;
  totalErrors: number;
}

export interface TestSummary {
  generatedAt: string;
  totalTests: number;
  coverage: CoverageStats;
  tests: Array<{
    replayId: string;
    duration: number;
    url: string;
    actionCount: number;
    errorCount: number;
  }>;
}

export class ReplayAnalyzer {
  private sessions: ReplaySession[] = [];

  /**
   * Analyze a single replay and extract actionable test data.
   */
  analyzeReplay(replayData: Record<string, any>): ReplaySession {
    const replayId = replayData.id || 'unknown';
    const duration = replayData.duration || 0;

    // Extract starting URL
    const url = this.extractUrl(replayData);

    // Parse events (if available in the replay data)
    const actions = this.extractActions(replayData);
    const networkRequests = this.extractNetworkRequests(replayData);
    const errors = this.extractErrors(replayData);

    const metadata: ReplayMetadata = {
      user: replayData.user,
      browser: replayData.browser,
      os: replayData.os,
      device: replayData.device,
      startedAt: replayData.started_at,
      finishedAt: replayData.finished_at,
      errorCount: replayData.error_count || 0,
      projectId: replayData.project_id,
    };

    const session: ReplaySession = {
      replayId,
      duration,
      url,
      actions,
      networkRequests,
      errors,
      metadata,
    };

    this.sessions.push(session);
    return session;
  }

  /**
   * Extract starting URL from replay data.
   */
  private extractUrl(replayData: Record<string, any>): string {
    // Check various possible fields
    if (replayData.urls && Array.isArray(replayData.urls) && replayData.urls.length > 0) {
      return replayData.urls[0];
    }
    if (replayData.url) {
      return replayData.url;
    }
    if (replayData.tags && Array.isArray(replayData.tags)) {
      for (const tag of replayData.tags) {
        if (tag.key === 'url') {
          return tag.value || '';
        }
      }
    }

    return 'https://example.com'; // Fallback
  }

  /**
   * Extract user actions from replay data.
   *
   * Note: This is a placeholder. Actual implementation depends on
   * Sentry's replay event format, which may require fetching
   * additional data via the events endpoint.
   */
  private extractActions(replayData: Record<string, any>): UserAction[] {
    const actions: UserAction[] = [];

    // Placeholder: In a real implementation, you'd parse the
    // rrweb events or Sentry's specific format
    // Example structure we might see:
    // {
    //   "type": 3,  // IncrementalSnapshot
    //   "data": {
    //     "source": 2,  // MouseInteraction
    //     "type": 2,  // Click
    //     "id": 123,
    //     "x": 100,
    //     "y": 200
    //   }
    // }

    return actions;
  }

  /**
   * Extract network requests from replay data.
   */
  private extractNetworkRequests(replayData: Record<string, any>): NetworkRequest[] {
    const requests: NetworkRequest[] = [];

    // Placeholder: Parse network activity from replay
    // Sentry may include this in breadcrumbs or separate events

    return requests;
  }

  /**
   * Extract errors from replay data.
   */
  private extractErrors(replayData: Record<string, any>): Array<Record<string, any>> {
    const errors: Array<Record<string, any>> = [];

    if (replayData.errors && Array.isArray(replayData.errors)) {
      return replayData.errors;
    } else if (replayData.error_ids && Array.isArray(replayData.error_ids)) {
      // Would need to fetch error details separately
      return replayData.error_ids.map((id: string) => ({ id }));
    }

    return errors;
  }

  /**
   * Analyze test coverage across all sessions.
   */
  analyzeCoverage(): CoverageStats {
    const allUrls: string[] = [];
    const allSelectors: string[] = [];
    const allBrowsers: string[] = [];

    for (const session of this.sessions) {
      allUrls.push(session.url);
      if (session.metadata.browser?.name) {
        allBrowsers.push(session.metadata.browser.name);
      }

      for (const action of session.actions) {
        if (action.selector) {
          allSelectors.push(action.selector);
        }
      }
    }

    // Count occurrences
    const urlDistribution = this.countOccurrences(allUrls);
    const browserDistribution = this.countOccurrences(allBrowsers);

    return {
      totalSessions: this.sessions.length,
      uniqueUrls: new Set(allUrls).size,
      uniqueSelectors: new Set(allSelectors).size,
      urlDistribution,
      browserDistribution,
      totalActions: this.sessions.reduce((sum, s) => sum + s.actions.length, 0),
      totalErrors: this.sessions.reduce((sum, s) => sum + s.errors.length, 0),
    };
  }

  /**
   * Count occurrences of items in an array.
   */
  private countOccurrences(items: string[]): Record<string, number> {
    const counts: Record<string, number> = {};
    for (const item of items) {
      counts[item] = (counts[item] || 0) + 1;
    }
    return counts;
  }

  /**
   * Select most valuable sessions for automated testing.
   */
  selectSessionsForTesting(options: {
    maxSessions?: number;
    prioritizeErrors?: boolean;
    minDuration?: number;
  } = {}): ReplaySession[] {
    const {
      maxSessions = 10,
      prioritizeErrors = true,
      minDuration = 1000, // milliseconds
    } = options;

    // Filter by minimum duration
    let candidates = this.sessions.filter((s) => s.duration >= minDuration);

    if (prioritizeErrors) {
      // Sort by error count (descending) then by duration
      candidates.sort((a, b) => {
        if (b.errors.length !== a.errors.length) {
          return b.errors.length - a.errors.length;
        }
        return b.duration - a.duration;
      });
    } else {
      // Sort by duration (longer sessions = more coverage)
      candidates.sort((a, b) => b.duration - a.duration);
    }

    return candidates.slice(0, maxSessions);
  }

  /**
   * Generate a Playwright test script from a session.
   */
  toPlaywrightTest(session: ReplaySession): string {
    const lines: string[] = [
      "import { test, expect } from '@playwright/test';",
      '',
      `test('replay-${session.replayId.substring(0, 8)}', async ({ page }) => {`,
      `  // Original session duration: ${session.duration}ms`,
      `  // URL: ${session.url}`,
      '',
      `  await page.goto('${session.url}');`,
      '',
    ];

    for (const action of session.actions) {
      if (action.actionType === 'click' && action.selector) {
        lines.push(`  await page.click('${action.selector}');`);
      } else if (action.actionType === 'input' && action.selector && action.value) {
        lines.push(`  await page.fill('${action.selector}', '${action.value}');`);
      } else if (action.actionType === 'navigation' && action.target) {
        lines.push(`  await page.goto('${action.target}');`);
      }

      // Add small delay to simulate real user behavior
      lines.push('  await page.waitForTimeout(100);');
      lines.push('');
    }

    lines.push('  // Take final screenshot for comparison');
    lines.push(`  await page.screenshot({ path: 'screenshots/replay-${session.replayId.substring(0, 8)}.png' });`);
    lines.push('});');

    return lines.join('\n');
  }

  /**
   * Export selected sessions as Playwright test scripts.
   */
  exportForTesting(outputDir: string = './tests'): void {
    // Create output directory
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    const selected = this.selectSessionsForTesting();

    for (const session of selected) {
      const testScript = this.toPlaywrightTest(session);
      const filename = path.join(outputDir, `replay-${session.replayId.substring(0, 8)}.spec.ts`);

      fs.writeFileSync(filename, testScript);
      console.log(`Generated test: ${filename}`);
    }

    // Generate a summary file
    const summary: TestSummary = {
      generatedAt: new Date().toISOString(),
      totalTests: selected.length,
      coverage: this.analyzeCoverage(),
      tests: selected.map((s) => ({
        replayId: s.replayId,
        duration: s.duration,
        url: s.url,
        actionCount: s.actions.length,
        errorCount: s.errors.length,
      })),
    };

    const summaryPath = path.join(outputDir, 'test-summary.json');
    fs.writeFileSync(summaryPath, JSON.stringify(summary, null, 2));

    console.log(`\nGenerated ${selected.length} tests in ${outputDir}/`);
    console.log(`Summary written to ${summaryPath}`);
  }

  /**
   * Get all analyzed sessions.
   */
  getSessions(): ReplaySession[] {
    return this.sessions;
  }
}

/**
 * Example usage of ReplayAnalyzer
 */
async function main() {
  // Example replay data (simplified)
  const sampleReplay = {
    id: 'abc123def456',
    duration: 45000, // 45 seconds
    url: 'https://example.com/app',
    user: { email: 'user@example.com' },
    browser: { name: 'Chrome', version: '120.0' },
    os: { name: 'macOS', version: '14.0' },
    error_count: 2,
    errors: [
      { message: 'TypeError: Cannot read property X' },
      { message: 'NetworkError: Failed to fetch' },
    ],
  };

  const analyzer = new ReplayAnalyzer();

  // Analyze replay
  const session = analyzer.analyzeReplay(sampleReplay);

  console.log('Analyzed Replay Session:');
  console.log(`  ID: ${session.replayId}`);
  console.log(`  Duration: ${session.duration}ms`);
  console.log(`  URL: ${session.url}`);
  console.log(`  Errors: ${session.errors.length}`);

  // Generate test script
  console.log('\nGenerated Test Script:');
  console.log(analyzer.toPlaywrightTest(session));

  // Coverage analysis
  console.log('\nCoverage Analysis:');
  const coverage = analyzer.analyzeCoverage();
  console.log(JSON.stringify(coverage, null, 2));
}

// Run if executed directly
if (require.main === module) {
  main();
}
